src/Microsoft.AspNetCore.Mvc.ViewFeatures/Internal/SaveTempDataPropertyFilterBase.cs

```diff
@@ -98,10 +98,18 @@ private static void ValidateProperty(PropertyHelper propertyHelper)
 
             var propertyType = Nullable.GetUnderlyingType(property.PropertyType) ?? property.PropertyType;
 
-            if (!(propertyType.GetTypeInfo().IsPrimitive || propertyType == typeof(string)))
+            if (!TempDataSerializer.CanSerializeType(propertyType, out var actualType, out var errorMessage))
             {
-                throw new InvalidOperationException(
-                    Resources.FormatTempDataProperties_PrimitiveTypeOrString(property.DeclaringType.FullName, property.Name, nameof(TempDataAttribute)));
+                var messageWithPropertyInfo = Resources.FormatTempDataProperties_InvalidType(
+                    property.DeclaringType.FullName,
+                    property.Name,
+                    nameof(TempDataAttribute));
+
+                errorMessage = errorMessage ?? Resources.FormatTempData_CannotSerializeType(
+                    typeof(TempDataSerializer).FullName,
+                    actualType);
+
+                throw new InvalidOperationException($"{messageWithPropertyInfo} {errorMessage}");
```

Comment:
The we we're building these error messages is just kind wacky. Use a stringbuilder with appendformat?